### PR TITLE
[SourceKit] Add accessibility in extension structure (SR-4823)

### PR DESCRIPTION
**Pull request by marcelofabri**
Adds the `key.accessibility` key in `source.lang.swift.decl.extension` structure.

Resolves [SR-4823](https://bugs.swift.org/browse/SR-4823).

### CODE REVIEW

**Comment by xwu** [comment]
...I may be misunderstanding what you're going for here, but as discussed previously on Swift-Evolution, extensions don't have their own accessibility. Sometimes, it is allowed to write `public extension`, for example, but that is intended to be a shorthand for all members being public. The extension itself continues to have no accessibility.

**Comment by marcelofabri** [comment]
I know that formally extensions themselves doesn't have accessibility, but since `public extension` is allowed, I think that information should be available in the structure for consumers.

For example, [SwiftLint](https://github.com/realm/SwiftLint) has a [rule](https://github.com/realm/SwiftLint/blob/master/Source/SwiftLintFramework/Rules/ExtensionAccessModifierRule.swift#L12) to enforce that if all the members from an extension have the same accessibility, you should use the shorthand form. Currently, the implementation needs to look for all tokens with `attributeBuiltin` kind, check if it's an ACL keyword and try to match with an extension declaration.

If this information was available on the extension structure, that implementation would be much simpler and reliable.

**Comment by xwu** [comment]
Cool, I can see how that's useful. However, it is misleading to label extensions as internal when they lack a modifier; that's simply not what they are.

**Comment by marcelofabri** [comment]
> it is misleading to label extensions as internal when they lack a modifier; that's simply not what they are.

What do you think it'd be better? Omit the key when there's no modifier?

**Comment by xwu** [comment]
IMO, that would be superior, yes.

**Comment by marcelofabri** [comment]
> Would that be a problematic approach?

From my point of view, no. I'd just like to hear from more people involved in SourceKit before implementing it. Maybe @nkcsgexi?

**Comment by nkcsgexi** [comment]
@marcelofabri  could you elaborated why we need to handle the accessibility of extension explicitly here? do we have issue with syntax highlighting?

**Comment by marcelofabri** [comment]
No issues with syntax highlighting, it's more about providing more information for clients. 

In particular, SwiftLint has some rules that would benefit from that information: 
* [**No Extension Access Modifier**](https://github.com/realm/SwiftLint/blob/master/Rules.md#no-extension-access-modifier): generates warnings if an accessibility is explicitly added to an extension.
* [**Extension Access Modifier**](https://github.com/realm/SwiftLint/blob/master/Rules.md#extension-access-modifier): generates a warning if all members of an extension have the accessibility (it could be only in the extension).
* [**Private over fileprivate**](https://github.com/realm/SwiftLint/blob/master/Rules.md#private-over-fileprivate): generates warnings when `fileprivate` is used in a place where `private` would be enough.

**Comment by nkcsgexi** [comment]
Interesting, thanks for clarification! Judging from the requirements, we really don't need to infer accessibility when an extension has no explicit one, am I right? i.e. we need to only report the accessibility user specified.

**Comment by marcelofabri** [comment]
Right, I just wanted to check if that'd be the preferred approach before making changes. 

Just to be clear: 
`key.accessibility` will only be present in `source.lang.swift.decl.extension` iff the accessibility was added explicitly.

**Comment by nkcsgexi** [comment]
Right, this seems to me a better approach. To do this, could you refactor the accessibility collection part of `inferAccessibility` to a new static function `getAccessibilityStrictly `, and using `getAccessibilityStrictly` here for your improvement?

**Comment by marcelofabri** [comment]
@nkcsgexi I've made the suggested changes in 86d9e91. 🎉

File: tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp [hunk 1]
```diff
@@ -912,13 +912,22 @@ static UIdent getAccessibilityUID(Accessibility Access) {
...
   if (auto *AA = ED->getAttrs().getAttribute<AccessibilityAttr>())
     return AA->getAccess();
 
+  return None;
+}
+
+static Accessibility inferDefaultAccessibility(const ExtensionDecl *ED) {
+  auto StrictAccess = getAccessibilityStrictly(ED);
+  if (StrictAccess.hasValue())
+    return StrictAccess.getValue();
```
Comment (nkcsgexi) [inline tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp]: Thanks! Minor nitpick: this above three lines can be combined to:
```swift
 if (auto StrictAccess = getAccessibilityStrictly(ED))
   return StrictAccess.getValue();
```

File: tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp [hunk 2]
```diff
@@ -1025,6 +1034,11 @@ class SwiftDocumentStructureWalker: public ide::SyntaxModelWalker {
     if (Node.Kind != SyntaxStructureKind::Parameter) {
       if (auto *VD = dyn_cast_or_null<ValueDecl>(Node.Dcl)) {
         AccessLevel = getAccessibilityUID(inferAccessibility(VD));
+      } else if (auto *ED = dyn_cast_or_null<ExtensionDecl>(Node.Dcl)) {
+        auto StrictAccess = getAccessibilityStrictly(ED);
+        if (StrictAccess.hasValue()) {
+          AccessLevel = getAccessibilityUID(StrictAccess.getValue());
+        }
```
Comment (nkcsgexi) [inline tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp]: Same here : )

**Review by nkcsgexi — APPROVED** [review]

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by marcelofabri** [comment]
🎉

**Pull request merged**